### PR TITLE
[REF] Cleanup CRM_Member_BAO_Membership::buildMembershipTypeValues

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Fixes function to be more readable, use caching & return more data

Before
----------------------------------------
- Function not using cached variables
- Function kinda long
- tax information not available
- No specific test cover

After
----------------------------------------
- MembershipTypes retrieved from cached function
- Additional tax info added to this cached function
- Various test bugs fixed & cover added

Technical Details
----------------------------------------
This is towards a fix for https://github.com/civicrm/civicrm-core/pull/15895 - I hit https://github.com/civicrm/civicrm-core/pull/16058 in the process. It doesn't fix #15895 but it points towards the solution which, to my mind, is assigning more useful values to the template & doing less in javascript. The next step is to use the correct tax_rate rather than have to use 2 arrays just to get the tax_rate - ie 

https://github.com/eileenmcnaughton/civicrm-core/commit/458a559a60e334b6897bec4542c594a9f34539c0#diff-fa0276456bc05f4629f775e6846f62ed

However, following that we can have the js use the figure which includes the tax, rather than having to calculate that in js with unknowns to deal with

Comments
----------------------------------------
_Anything else you would like the reviewer to note_

### CODE REVIEW

File: tests/phpunit/CRM/Member/BAO/MembershipTest.php [hunk 1]
```diff
@@ -650,21 +654,14 @@ public function testStaleMembership() {
...
-    $config = CRM_Core_Config::singleton();
-    $config->keyDisable = TRUE;
-
-    $membershipRenewal = new CRM_Core_Form();
-    $membershipRenewal->controller = new CRM_Core_Controller();
     list($MembershipRenew) = CRM_Member_BAO_Membership::processMembership(
       $contactId,
       $this->_membershipTypeID,
       FALSE,
-      $membershipRenewal,
```
Comment (eileenmcnaughton) [inline tests/phpunit/CRM/Member/BAO/MembershipTest.php]: This param is a bool!

File: tests/phpunit/CRM/Member/BAO/MembershipTest.php [hunk 2]
```diff
@@ -811,7 +808,7 @@ public function testMembershipPaymentForSingleContributionMultipleMembership() {
     $parentContactId = $this->individualCreate();
```
Comment (eileenmcnaughton) [inline tests/phpunit/CRM/Member/BAO/MembershipTest.php]: This test actually didn't pass in isolation for set-up related reasons

File: tests/phpunit/CRM/Member/BAO/MembershipTest.php [hunk 3]
```diff
@@ -871,67 +846,105 @@ public function testMembershipPaymentForSingleContributionMultipleMembership() {
...
+    $this->validateAllPayments();
+    $this->validateAllContributions();
+  }
+
+  /**
+   * Test the buildMembershipTypeValues function.
+   *
+   * @throws \CiviCRM_API3_Exception
+   */
+  public function testBuildMembershipTypeValues() {
```
Comment (eileenmcnaughton) [inline tests/phpunit/CRM/Member/BAO/MembershipTest.php]: here is the new testt

**Comment by seamuslee001** [comment]
@eileenmcnaughton 2 of the test failure look like they relate

**Comment by eileenmcnaughton** [comment]
@seamuslee001 this has passed now

**Comment by mattwire** [comment]
Useful step in the right direction

**Pull request merged**